### ISSUE TITLE
Creating a fork to replace comments models with one based on treebeard

### ISSUE THREAD

**Issue by kutenai** [danirus/django-comments-xtd#123]
I've spent some reviewing the source of this package, the original django-contrib-comments package, and any others I could find related to commenting (there are many). This package is the closest to what I would **like** to see, but I'm really looking for a comment system that uses a __modern__ tree approach, specifically django-treebeard.

Some issues I have with the current appraoch:
- each comment contains a Generic FK to the "connected" model
- tree traversal is done using "other" methods, i.e. not optimized for a tree structure

Given the limitations, I don't see a way around forking, renaming, and applying some pretty major internal changes. But, I like the core, and plan to just adapt it.

I'm posting this as information, but also as a "Hey, here is what I'm doing...", and asking the following questions:
- Does anyone see issues with what I'm doing? 
- Does this exist somewhere already I just didn't find it? 
- Would others find this useful also?

I intent to rename this as 'dango-comments-tree' (treebeard seemed too long).

I also have good React experience, and want to re-work that part of the app as well, probably some major re-writes of that.

One final (probably controversial) note -- I'm doing this for a project of mine, so, I'm time crunched. I intent to make this Django 2.0+ only, Python 3.5+ only (or maybe even 3.7+ - I really like F' strings).  

Comments welcome

**Comment by danirus** [comment]
Sounds as an interesting exercise.  
This app is attached to django-contrib-comments, as it extends it. What you propose, as you well said, requires a new model layer. Support for nested comments would benefit from it. In particular the `tree_from_queryset` method in the models.py module, that builds the comment tree out of the queryset.

How would you refer to the "connected" model if not using a Generic FK?

**Comment by kutenai** [comment]
I do not see another way to handle the foreign key relationship, so the Generic FK is still the best solution there, I just wanted to use a different approach to managing hierarchical relationships.

I've also been looking at a way to get user updates. One is to use websockets, probably with django channels. I'm thinking that is too much to work out. The other idea I had is to use an alternate notification approach for the client, specifically using firebase. 

New comments could be added via an API call, which would insert the comment into the DB, and update a firebase backend. Clients would register with firebase and be notified of any updates. They would then query the backend API to get the updated values. 

This approach would keep all of the data in Django, but use firebase as a notification mechanism. Other mechanisms like that could work also.

Frankly, this is all quite a bit bigger than I had anticipated, and my project is way behind. I need to just hack something together and make it work. This library is the closest to what I need, but I'm afraid I'll have to hack it up quite a bit to make it work, and so maintaining a mergeable version will be tough.

**Comment by kutenai** [comment]
Okay -- let me try this again. I had to go back and review what I did, since I have not touched it for a few weeks.

My plan is to associate a set of comments on a particular model using a `CommentAssociation` model. This model uses a Generic FK to the Object being associated but associates the "root" of the comment stream. This is quite a bit different than the existing method which associates each and every comment with a model using the GFK.

I re-worked the models to do this, but then I was working my way through some of your other code and found a roadblock. You have existing code that queries a list of comments for a queryset. With my new model approach, that isn't possible (or practical). So, that actually caused me to set this aside for a bit, as I realized I was in for a bigger project. 

At this point I'm going to have to re-work a lot of the internals, so it's not going to be a fork that can be re-merged into the source branch.

### Real-time comment updates

I've also been looking at a way to get real-time comment updates. One approach is to use websockets, probably with Django channels, but I don't want to go that rounte right now. I'm thinking that is too much to work out. 

The other idea I had is to use an alternate notification approach for the client, specifically using firebase.

New comments could be added via an API call, which would insert the comment into the DB, and update a firebase backend. Clients would register with firebase and be notified of any updates. They would then query the backend API to get the updated values.

This approach would keep all of the data in Django, but use firebase as a notification mechanism. 

Frankly, this is all quite a bit bigger than I had anticipated, and my project is way behind. I need to just hack something together and make it work. This library is the closest to what I need, but I'm afraid I'll have to hack it up quite a bit to make it work, and so maintaining a mergeable version will be tough.

**Comment by kutenai** [comment]
@danirus - I have made progress on my fork of this repo, so I thought I'd let you know.
The changes are too large to submit a PR, and I really just wanted to move forward with the new design, and skip any sort of backward compatibility.
The repo is here: https://github.com/sharpertool/django-comments-tree

I dropped 2.x support as well, as I'm all in on 3.6+, and don't want to look back.

The library has greatly reduced the dependency on django-comments, and I'll probably just drop that at some point. There were too many issues with the way the old module tried to work, so I've copied most stuff into the _tree version.

As I said before, this version uses django-treebeard with the MP_Node to have efficient handling of hierarchy.
It uses an external class `CommentAssociation` that connects a comment root to a generic object. This is all pretty efficient.

I also use django-markupfield in order to have a more robust format for the comments. Markup field allows the comments to be stored in a configuration format (markdown, or something else), and it stores the raw data, as well as a pre-rendered version. This allows the comments to be rendered quicly, but edited in a rich text editor.

Anyway,  hit me up if you want to chat more.
Cheers,
Ed
